File: 2005/Windows SDK NetFx 3.0/Technologies/Localization/StringWalker/CS/StringWalker/StringWalker.cs
//-----------------------------------------------------------------------
//  This file is part of the Microsoft .NET Framework SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//-----------------------------------------------------------------------
// --------------------------------------------------------------------------
//  StringWalker class: this string encapsulates string operations in order
//  to provide easy methods that handle surrogates and combining characters
// --------------------------------------------------------------------------

using System;
using System.Globalization;
using System.Text;
using System.Runtime.Serialization;

namespace Microsoft.Samples.StringWalker
{
	[Serializable]
	public class StringWalkerException : Exception
	{
		public StringWalkerException()           : base() { }
		public StringWalkerException(String message) : base(message) { }
		public StringWalkerException(String message, Exception innerException) : base(message, innerException) { }
		protected StringWalkerException(SerializationInfo info, StreamingContext context) : base(info, context) { }
	}

	public class StringWalker
	{
		// private members
		private string myString = String.Empty;

		private int[] myIndex;

		private int myPos;

		// public property
		public int Length
		{
			get
			{
				return (null != myIndex) ? myIndex.Length : 0;
			}
		}

		// constructor
		public StringWalker(string input)
		{
			Initialize(input);
		}

		// ToString overriden method
		public override string ToString()
		{
			return myString;
		}

		// "easy" walking methods: GetFirst, GetNext, GetPrev, GetLast, Get
		// these are basically wrapper around the GetSubString method
		public bool GetFirst(out string input)
		{
			myPos = 0;
			return Get(myPos, out input);
		}

		public bool GetLast(out string input)
		{
			myPos = Length - 1;
			return Get(myPos, out input);
		}

		public bool GetNext(out string input)
		{
			return Get(++myPos, out input);
		}

		public bool GetPrevious(out string input)
		{
			return Get(--myPos, out input);
		}

		public bool Get(int index, out string input)
		{
			return (0 != GetSubString(index, 1, out input));
		}

		// GetSubString method
		public int GetSubString(int index, int count, out string input)
		{
			// check for index within bounds and non zero count
			if ((1 <= count) && (0 <= index) && (index < Length))
			{
				try
				{
					int lastindex = index + count;

					// if we are past the last char, then we get the string
					// up to the last char and return the actual count
					if (lastindex > (Length - 1))
					{
						input = myString.Substring(myIndex[index]);
						return Length - index;
					}
					else
					{
						input = myString.Substring(myIndex[index], myIndex[lastindex] - myIndex[index]);
						return count;
					}
				}
				catch // catch all and throw our exception
				{
					throw (new StringWalkerException());
				}
			}
			else
			{
				input = String.Empty;
				return 0;
			}
		}

		// Insert, Remove: both methods return true if the operation was succesful and false otherwise
		// Insert: inserts a string at the specified position
		public bool Insert(int index, string input)
		{
			if ((0 <= index) && (index <= Length))
			{
				try
				{
					if (index == Length)
						Initialize(myString.Insert(myString.Length, input));
					else
						Initialize(myString.Insert(myIndex[index], input));

					return true;
				}
				catch // catch all and throw our exception
				{
					throw (new StringWalkerException());
				}
			}

			return false;
		}

		// Remove: removes the specified number of text elements starting at the specified position
		public bool Remove(int index, int count)
		{
			if ((count > 0) && (0 <= index) && (index < Length))
			{
				try
				{
					int idxLast = index + count;
					int charcount = (idxLast < Length) ? myIndex[idxLast] - myIndex[index] : myString.Length - myIndex[index];

					Initialize(myString.Remove(myIndex[index], charcount));
					return true;
				}
				catch // catch all and throw our exception
				{
					throw (new StringWalkerException());
				}
			}

			return false;
		}

		// IndexOf: 
		public int IndexOf(string input, int index)
		{
			if ((0 <= index) && (index < Length))
			{
				try
				{
					// try and find the input string in the current string
					int position = myIndex[index];
					int foundAt = myString.IndexOf(input, position);

					// if the string is found, then we need to see if it
					// can be matched to a text element index.
					if (-1 != foundAt)
					{
						for (int i = 0; i < myIndex.Length; i++)
							if (myIndex[i] == foundAt)
								return i;
					}
				}
				catch // catch all and throw our exception
				{
					throw (new StringWalkerException());
				}
			}

			return -1;
		}

		// private initialization method
		private void Initialize(string input)
		{
			try
			{
				myPos = 0;
				myString = input;
				myIndex = StringInfo.ParseCombiningCharacters(myString);
			}
			catch (ArgumentNullException)
			{
				throw (new StringWalkerException());
			}
		}
	}
}

File: 2005/Windows SDK NetFx 3.0/Technologies/Localization/StringWalker/VB/StringWalker/StringWalker.vb
'-----------------------------------------------------------------------
'  This file is part of the Microsoft .NET Framework SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'-----------------------------------------------------------------------
'  StringWalker class: this string encapsulates string operations in order
'  to provide easy methods that handle surrogates and combining characters
' --------------------------------------------------------------------------

Imports System
Imports System.Runtime.Serialization
Imports System.Globalization
Imports System.Text

Namespace Microsoft.Samples.StringWalker
    <Serializable()> _
    Public Class StringWalkerException
        Inherits Exception

        Public Sub New()
            MyBase.New()
        End Sub

        Public Sub New(ByVal message As String)
            MyBase.New(message)
        End Sub


        Public Sub New(ByVal message As String, ByVal innerException As Exception)
            MyBase.New(message, innerException)
        End Sub

        Protected Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)
            MyBase.New(info, context)
        End Sub

    End Class

    Public Class StringWalker
        ' private members
        Private myString As String = String.Empty
        Private myIndex() As Integer
        Private myPos As Integer

        ' public property

        Public ReadOnly Property Length() As Integer
            Get
                If myIndex Is Nothing Then
                    Return 0
                Else
                    Return myIndex.Length
                End If
            End Get
        End Property

        ' constructor
        Public Sub New(ByVal input As String)
            Initialize(input)
        End Sub 'New

        ' ToString overriden method
        Public Overrides Function ToString() As String
            Return myString
        End Function 'ToString

        ' "easy" walking methods: GetFirst, GetNext, GetPrev, GetLast, Get
        ' these are basically wrapper around the GetSubString method
        Public Function GetFirst(ByRef input As String) As Boolean
            myPos = 0
            Return GetElement(myPos, input)
        End Function 'GetFirst

        Public Function GetLast(ByRef input As String) As Boolean
            myPos = Length - 1
            Return GetElement(myPos, input)
        End Function 'GetLast

        Public Function GetNext(ByRef input As String) As Boolean
            myPos = myPos + 1
            Return GetElement(myPos, input)
        End Function 'GetNext

        Public Function GetPrevious(ByRef input As String) As Boolean
            myPos = myPos - 1
            Return GetElement(myPos, input)
        End Function 'GetPrev

        Public Function GetElement(ByVal index As Integer, ByRef input As String) As Boolean
            Return 0 <> GetSubString(index, 1, input)
        End Function 'Get

        ' GetSubString method
        Public Function GetSubString(ByVal index As Integer, ByVal count As Integer, ByRef input As String) As Integer
            ' check for index within bounds and non zero count
            If 1 <= count And 0 <= index And index < Length Then
                Try
                    Dim lastindex As Integer = index + count

                    ' if we are past the last char, then we get the string
                    ' up to the last char and return the actual count
                    If lastindex > Length - 1 Then
                        input = myString.Substring(myIndex(index))
                        Return Length - index
                    Else
                        input = myString.Substring(myIndex(index), myIndex(lastindex) - myIndex(index))
                        Return count
                    End If
                Catch ' catch all and throw our exception
                    Throw (New StringWalkerException())
                End Try
            End If
            input = String.Empty
            Return 0
        End Function 'GetSubString

        ' Insert, Remove: both methods return true if the operation was succesful and false otherwise
        ' Insert: inserts a string at the specified position
        Public Function Insert(ByVal index As Integer, ByVal input As String) As Boolean
            If 0 <= index And index <= Length Then
                Try
                    If index = Length Then
                        Initialize(myString.Insert(myString.Length, input))
                    Else
                        Initialize(myString.Insert(myIndex(index), input))
                    End If
                    Return True
                Catch ' catch all and throw our exception
                    Throw (New StringWalkerException())
                End Try
            End If
            Return False
        End Function 'Insert

        ' Remove: removes the specified number of text elements starting at the specified position
        Public Function Remove(ByVal index As Integer, ByVal count As Integer) As Boolean
            If count > 0 And 0 <= index And index < Length Then
                Try
                    Dim idxLast As Integer = index + count
                    Dim charcount As Integer

                    If idxLast < Length Then
                        charcount = myIndex(idxLast) - myIndex(index)
                    Else
                        charcount = myString.Length - myIndex(index)
                    End If

                    Initialize(myString.Remove(myIndex(index), charcount))
                    Return True
                Catch ' catch all and throw our exception
                    Throw (New StringWalkerException())
                End Try
            End If
            Return False
        End Function 'Remove

        ' IndexOf: 
        Public Function IndexOf(ByVal input As String, ByVal index As Integer) As Integer
            If 0 <= index And index < Length Then
                Try
                    ' try and find the input string in the current string
                    Dim position As Integer = myIndex(index)
                    Dim foundAt As Integer = myString.IndexOf(input, position)

                    ' if the string is found, then we need to see if it
                    ' can be matched to a text element index.
                    If -1 <> foundAt Then
                        Dim i As Integer
                        For i = 0 To myIndex.Length - 1
                            If myIndex(i) = foundAt Then
                                Return i
                            End If
                        Next i
                    End If
                Catch ' catch all and throw our exception
                    Throw (New StringWalkerException())
                End Try
            End If
            Return -1
        End Function 'IndexOf

        ' private initialization method
        Private Sub Initialize(ByVal input As String)
            Try
                myPos = 0
                myString = input
                myIndex = StringInfo.ParseCombiningCharacters(myString)

            Catch ' catch all and throw our exception
                Throw (New StringWalkerException())
            End Try
        End Sub 'Initialize

    End Class 'StringWalker
End Namespace